File: local_agent/telegram_monitor.py
"""
🍯 Honey-Pot — Telegram Local Monitor
=======================================
Runs silently in background. Monitors YOUR Telegram account for incoming messages.
When a scam is detected, the bot auto-replies on YOUR behalf.

Uses Telethon (Telegram MTProto User API — acts AS you, not as a bot).

Requirements:
    pip install telethon requests python-dotenv

Setup:
    1. Get API credentials from https://my.telegram.org/apps
    2. Add to .env:  TG_API_ID, TG_API_HASH, HONEYPOT_API_KEY
    3. Run: python telegram_monitor.py
    4. First run will ask for your phone number + OTP (one-time login)
"""

import os
import logging
import asyncio
import requests
from telethon import TelegramClient, events
from dotenv import load_dotenv

from session_cache import add_message, get_history, clear_history

load_dotenv()

# ─── Config ────────────────────────────────────────────────────────────────────
TG_API_ID       = int(os.getenv("TG_API_ID", "0"))
TG_API_HASH     = os.getenv("TG_API_HASH", "")
HONEYPOT_URL    = os.getenv("HONEYPOT_URL", "https://honey-pot-agent.onrender.com/honeypot/message")
HONEYPOT_KEY    = os.getenv("HONEYPOT_API_KEY", "")
SESSION_NAME    = "honeypot_monitor"   # Local session file name

# Chats to IGNORE auto-reply (add your groups/channels/trusted contacts here)
IGNORED_SENDERS = set()  # e.g. {123456789, 987654321}  ← Telegram user IDs

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%H:%M:%S"
)
log = logging.getLogger("TelegramMonitor")

# ─── Honey-Pot API Call ─────────────────────────────────────────────────────────
def check_message(session_id: str, text: str) -> str | None:
    """
    Send message to Honey-Pot API with full conversation history.
    Returns the agent reply if scam detected, None if safe.
    """
    try:
        add_message(session_id, "scammer", text)
        history = get_history(session_id)
        resp = requests.post(
            HONEYPOT_URL,
            json={
                "sessionId": session_id,
                "message": {
                    "sender": "scammer",
                    "text": text
                },
                "conversationHistory": history[:-1],  # exclude current msg
                "metadata": {"channel": "Telegram"}
            },
            headers={"x-api-key": HONEYPOT_KEY},
            timeout=10
        )
        resp.raise_for_status()
        data = resp.json()
        reply = data.get("reply")
        if reply:
            add_message(session_id, "agent", reply)
        return reply
    except Exception as e:
        log.error(f"API error: {e}")
        return None  # On error, do NOT auto-reply (safe default)


# ─── Main Monitor ───────────────────────────────────────────────────────────────
async def main():
    if not TG_API_ID or not TG_API_HASH:
        log.error("[FAIL] Missing TG_API_ID / TG_API_HASH in .env")
        return
    if not HONEYPOT_KEY:
        log.error("[FAIL] Missing HONEYPOT_API_KEY in .env")
        return

    client = TelegramClient(SESSION_NAME, TG_API_ID, TG_API_HASH)
    await client.start()   # First run: will ask for phone + OTP

    me = await client.get_me()
    log.info(f"[OK] Logged in as: {me.first_name} (@{me.username})")
    log.info("[ACTIVE] Honey-Pot monitor active — watching for scams silently...")

    @client.on(events.NewMessage(incoming=True))
    async def handler(event):
        # Only monitor private (1-on-1) messages, not group chats
        if not event.is_private:
            return

        sender = await event.get_sender()
        sender_id = sender.id
        text = event.raw_text.strip()

        # Skip empty messages and whitelisted senders
        if not text or sender_id in IGNORED_SENDERS:
            return

        session_id = f"tg_{sender_id}"
        log.info(f"📩 Message from {getattr(sender, 'first_name', sender_id)}: {text[:60]}...")

        # Call Honey-Pot API
        reply = check_message(session_id, text)

        if reply:
            log.warning(f"[SCAM] SCAM DETECTED — auto-replying on your behalf")
            await event.reply(reply)   # Sends as YOU, not as a bot
            log.info(f"[OK] Reply sent: {reply[:80]}...")
        else:
            log.info(f"[OK] Safe message — no action taken")

    await client.run_until_disconnected()


if __name__ == "__main__":
    asyncio.run(main())

File: local_agent/telegram_monitor.js
/**
 * 🍯 Honey-Pot — Telegram Local Monitor
 * ========================================
 * Runs silently in background. Monitors YOUR Telegram bot.
 * When a scam is detected, auto-replies on behalf of the bot.
 *
 * Uses node-telegram-bot-api (Telegram Bot API).
 *
 * Requirements:
 *   npm install node-telegram-bot-api axios dotenv
 *
 * Setup:
 *   1. Create a bot via @BotFather on Telegram
 *   2. Get your bot token (format: 123456789:ABCdefGHIjklMNOpqrsTUVwxyz)
 *   3. Run: node telegram_monitor.js <BOT_TOKEN>
 *   4. Done — it runs silently monitoring for scam messages
 */

require("dotenv").config();
const TelegramBot = require("node-telegram-bot-api");
const axios = require("axios");

// ─── Config ────────────────────────────────────────────────────────────────────
// Get bot token from command line argument
const TELEGRAM_BOT_TOKEN = process.argv[2] || process.env.TELEGRAM_BOT_TOKEN || "";
const HONEYPOT_URL = process.env.HONEYPOT_URL || "https://honey-pot-agent.onrender.com/honeypot/message";
const HONEYPOT_KEY = process.env.HONEYPOT_API_KEY || "";

// User IDs to NEVER auto-reply to (add trusted contacts here)
const WHITELISTED = new Set([]);

// In-memory conversation history per session
const conversationHistory = {};

// ─── Telegram Bot Setup ─────────────────────────────────────────────────────────
if (!TELEGRAM_BOT_TOKEN) {
    console.error("❌ TELEGRAM_BOT_TOKEN not set in .env");
    console.error("   Get your token from @BotFather on Telegram");
    process.exit(1);
}

if (!HONEYPOT_KEY) {
    console.error("❌ HONEYPOT_API_KEY not set in .env");
    process.exit(1);
}

const bot = new TelegramBot(TELEGRAM_BOT_TOKEN, { polling: true });

console.log("🍯 Starting Honey-Pot Telegram Monitor...");
console.log("👁️  Watching all incoming messages silently...\n");

// ─── Conversation History Helpers ──────────────────────────────────────────────
function addHistory(sessionId, sender, text) {
    if (!conversationHistory[sessionId]) {
        conversationHistory[sessionId] = [];
    }
    conversationHistory[sessionId].push({ sender, text });
    // Keep last 20 messages
    if (conversationHistory[sessionId].length > 20) {
        conversationHistory[sessionId] = conversationHistory[sessionId].slice(-20);
    }
}

function getHistory(sessionId) {
    return conversationHistory[sessionId] || [];
}

// ─── Honey-Pot API Call ─────────────────────────────────────────────────────────
async function checkMessage(sessionId, text) {
    try {
        addHistory(sessionId, "scammer", text);
        const history = getHistory(sessionId).slice(0, -1); // exclude current
        const response = await axios.post(
            HONEYPOT_URL,
            {
                sessionId: sessionId,
                message: { sender: "scammer", text: text },
                conversationHistory: history,
                metadata: { channel: "Telegram" }
            },
            {
                headers: {
                    "x-api-key": HONEYPOT_KEY,
                    "Content-Type": "application/json"
                },
                timeout: 10000
            }
        );
        const reply = response.data.reply || null;
        if (reply) {
            addHistory(sessionId, "agent", reply);
        }
        return reply;
    } catch (err) {
        console.error("⚠️  API error:", err.message);
        return null;  // On error, do NOT auto-reply (safe default)
    }
}

// ─── Message Handler ───────────────────────────────────────────────────────────
bot.on('message', async (msg) => {
    // Skip messages from the bot itself
    if (msg.from.id === bot.id) return;
    
    // Skip group chats (only monitor private messages)
    if (msg.chat.type !== 'private') return;

    const userId = msg.from.id.toString();
    const text = msg.text?.trim();

    if (!text || WHITELISTED.has(userId)) return;

    const sessionId = `tg_${userId}`;
    const username = msg.from.username ? `@${msg.from.username}` : `User ${userId}`;
    
    console.log(`📩 Message from ${username} (${userId}): ${text.substring(0, 60)}...`);

    const reply = await checkMessage(sessionId, text);

    if (reply) {
        console.warn(`⚠️  SCAM DETECTED — auto-replying as bot`);
        await bot.sendMessage(msg.chat.id, reply);
        console.log(`✅ Reply sent: ${reply.substring(0, 80)}...`);
    } else {
        console.log(`✅ Safe message — no action taken`);
    }
});

// ─── Error Handling ─────────────────────────────────────────────────────────────
bot.on('polling_error', (error) => {
    console.error('❌ Polling error:', error.message);
});

process.on('uncaughtException', (err) => {
    console.error('❌ Uncaught Exception:', err.message);
});

process.on('SIGINT', () => {
    console.log('\n🍯 Stopping Telegram Monitor...');
    bot.stopPolling();
    process.exit(0);
});

console.log('✅ Telegram Monitor started successfully');
console.log('📱 Bot is now listening for messages...\n');
